### PR TITLE
Added sale payment

### PR DESCRIPTION
**Pull request by prakashpp**

### CODE REVIEW

File: invoice.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    def auto_pay_from(self, sale):
+        """
+        Automatically try to pay the invoice from given sale.
+        If there are open authorizations, use them, if not start a new charge.
+        """
+        PaymentTransaction = Pool().get('payment_gateway.transaction')
+
+        amount = self.amount_to_pay_today
+
+        method_order = ['manual', 'credit_card']
```
Comment (sharoonthomas) [inline invoice.py:26]: this block is a repeating pattern. I think this should be part of the sale payment sort order itself. This allows other modules (like gift card ) to introduce payment processing into the picture.

File: payment.py [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+from trytond.pool import Pool
+
+__all__ = ['SalePayment']
+
+
+class SalePayment(ModelSQL, ModelView):
+    "Sale Payment"
+    __name__ = 'sale.payment'
+
+    sequence = fields.Integer('Sequence', required=True)
```
Comment (sharoonthomas) [inline payment.py:21]: select=True

File: payment.py [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+        fields.Many2One('party.party', 'Party'),
+        'on_change_with_party'
+    )
+    currency_digits = fields.Function(
+        fields.Integer('Currency Digits'),
+        'on_change_with_currency_digits'
+    )
+    gateway = fields.Many2One(
+        'payment_gateway.gateway', 'Gateway', required=True,
+        ondelete='RESTRICT',
```
Comment (sharoonthomas) [inline payment.py:33]: select

File: payment.py [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+    )
+    amount_remaining = fields.Function(
+        fields.Numeric(
+            'Amount Remaining', digits=(16, Eval('currency_digits', 2)),
+            depends=['currency_digits'],
+        ), 'get_amount_remaining'
+    )
+    payment_transactions = fields.One2Many(
+        'payment_gateway.transaction', 'payment', 'Payment Transactions',
+    )
```
Comment (sharoonthomas) [inline payment.py:70]: readonly

File: payment.py [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+        }
+    )
+
+    @staticmethod
+    def default_sequence():
+        return 10
+
+    @fields.depends('sale')
+    def on_change_with_currency_digits(self, name=None):
+        if self.sale.currency:
```
Comment (sharoonthomas) [inline payment.py:83]: `if self.sale and self.sale.currency`

File: payment.py [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+            return self.sale.currency.digits
+        return 2
+
+    @fields.depends('sale')
+    def on_change_with_party(self, name=None):
+        return self.sale.party.id
```
Comment (sharoonthomas) [inline payment.py:89]: check if sale exists first

File: payment.py [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+        """
+        PaymentTransaction = Pool().get('payment_gateway.transaction')
+
+        payment_transactions = PaymentTransaction.search([
+            ('payment', '=', self)
+        ])
+        amount_consumed = Decimal('0')
+        for transaction in payment_transactions:
+            if transaction.state not in ('authorized', 'completed', 'posted'):
+                continue    # pragma: no cover
```
Comment (sharoonthomas) [inline payment.py:116]: needs coverage. Use a dummy gateway and use a cancelled transaction.

File: payment.py [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+    def authorize(self, amount, description):
+        """
+        Authorize the given amount from this transaction
+        """
+        PaymentTransaction = Pool().get('payment_gateway.transaction')
+        Date = Pool().get('ir.date')
+
+        if amount > self.amount_remaining:
+            self.raise_user_error(
+                'Insufficient amount remaining in payment\n'
```
Comment (sharoonthomas) [inline payment.py:149]: move error to _error_messages

File: payment.py [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+                    len(self.payment_transactions),
+                )
+            )
+
+        if self.method == 'credit_card' and self.sale.last_card_payment != self:
+            amount = min(amount, self.amount_remaining)
+
+        transaction, = PaymentTransaction.create([{
+            'description': description or 'Auto charge from payment',
+            # 'origin': self,
```
Comment (sharoonthomas) [inline payment.py:166]: re-add origin

File: payment.py [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+                    self.amount_remaining,
+                    len(self.payment_transactions),
+                )
+            )
+
+        if self.method == 'credit_card' and self.sale.last_card_payment != self:
+            amount = min(amount, self.amount_remaining)
+
+        transaction, = PaymentTransaction.create([{
+            'description': description or 'Auto charge from payment',
```
Comment (sharoonthomas) [inline payment.py:165]: Auto charge from Sale

File: payment.py [hunk 11]
```diff
@@ -0,0 +1,194 @@
...
+
+    @classmethod
+    def delete(cls, payments):
+        """
+        Delete a payment only if there is no amount consumed
+        """
+        for payment in payments:
+            if payment.amount_consumed:
+                cls.raise_user_error(
+                    "This Payment cannot be deleted as some amount has " +
```
Comment (sharoonthomas) [inline payment.py:191]: move error to `_error_messages`

File: sale.py [hunk 12]
```diff
@@ -0,0 +1,402 @@
...
+    payment_remaining = fields.Function(
+        fields.Numeric(
+            'Payment Remaining', digits=(16, Eval('currency_digits', 2)),
+            depends=['currency_digits'],
+            help="Total value of payments remaining."
+        ), 'get_payment_amounts',
+    )
+    amount_invoiced = fields.Function(
+        fields.Numeric(
+            'Amount Invoices', digits=(16, Eval('currency_digits', 2)),
```
Comment (sharoonthomas) [inline sale.py:44]: Amount Invoiced

File: sale.py [hunk 13]
```diff
@@ -0,0 +1,402 @@
...
+    def authorize_from_sale_payments(self, amount, description):
+        """
+        Authorize the given amount from available sale payments
+
+        :param amount: Decimal amount in sale currency to capture.
+        :param description: Description for the payment transaction.
+        """
+        if amount > self.payment_remaining:
+            self.raise_user_error(
+                'Insufficient amount remaining in sale payments\n'
```
Comment (sharoonthomas) [inline sale.py:143]: move message to `_error_messages`

File: sale.py [hunk 14]
```diff
@@ -0,0 +1,402 @@
...
+                'Payments: %s' % (
+                    amount,
+                    self.payment_total,
+                    self.payment_remaining,
+                    len(self.payments),
+                )
+            )
+
+        transactions = []
+        order = ['manual', 'credit_card']
```
Comment (sharoonthomas) [inline sale.py:156]: this sorting as i said before should be in the tenders default order

File: sale.xml [hunk 15]
```diff
@@ -0,0 +1,20 @@
...
+        <record model="ir.action.wizard" id="wizard_add_payment">
+            <field name="name">Add Payment</field>
+            <field name="wiz_name">sale.payment.add</field>
+            <field name="model">sale.sale</field>
+        </record>
+        <record model="ir.ui.view" id="sale_payment_add_view_form">
+            <field name="model">sale.payment.add_view</field>
+            <field name="type">form</field>
+            <field name="name">sale_payment_add_form</field>
+        </record>
```
Comment (sharoonthomas) [inline sale.xml:18]: There should be a view of sale payment tree and form

**Comment by sharoonthomas** [comment]
Also add permissions:
1. Only sale admin can edit and delete sale payments
2. Only sale users can add and view payments

**Comment by prakashpp** [comment]
This is already merged to develop

**Pull request closed without merging**